feat(ngTransclude & $compile): fix treatment of unfilled optional slots

petebacondarwin opened the pull request:
Closes #13426

petebacondarwin commented:
@PascalPrecht - hey look at this!

PascalPrecht commented:
@petebacondarwin awesome! Does this also keep the original contents of the DOM where we transclude into?

E.g.
```
<span ng-transclude="fooSlot">This is default DOM and going to be replaced if `fooSlot` is defined, if not this stays here</span>
```

petebacondarwin commented:
Yep! That is the second commit in the PR.

PascalPrecht commented:
Argh. I just've just read the commit messages instead of reading the code haha. Awesome! Thanks for the work!

petebacondarwin commented:
See https://ci.angularjs.org/job/angular.js-pete/989/artifact/build/docs/api/ng/directive/ngTransclude#multi-slot-transclusion
for a running example

Narretz commented on src/ng/compile.js:
```diff
@@ -2222,13 +2235,15 @@ function $CompileProvider($provide, $$sanitizeUriProvider) {
+            if (slotTranscludeFn) {
+              return slotTranscludeFn(scope, cloneAttachFn, transcludeControllers, futureParentElement, scopeToChild);
+            }
+            if (isUndefined(slotTranscludeFn)) {
```
how about just `else`?

petebacondarwin replied:
We have to differentiate between `null` and `undefined`. I will add a comment to clarify.

Narretz replied:
Okay, I just noticed there's no difference in test behavior.

Narretz commented:
`feat($compile): expose transclusion `$$slots` on the `$transclude`
Isn't the commit message a bit contradictory? It's called a feature but `$$slots` indicates a private API.

And a question, is it possible to keep the content with the default behavior (no slots, transclude is true or content)? Since people have requested something similar for the default ngTransclude: https://github.com/angular/angular.js/pull/11839
I think we could extend the 'content' value to make it optional by adding `?` and applying the same behavior.

petebacondarwin commented:
We could make `$$slots` -> `$slots`.

petebacondarwin commented:
Regarding default content for the default transclusion, perhaps we could special case `""` in the transclude map:

```js
transclude: {
  'required': 'requiredSlot',
  'optional': '?optionalSlot',
  '': '?' // this would make the default transclusion optional too...
}
```

Narretz commented:
Then we should also document `$slots`, I guess.

For default transchlusion. Okay, I forgot you can have default transclusion and slots. The proposed API for that looks a bit silly, but it's probably okay. But we can also decide on that after 1.5

petebacondarwin committed:
feat(ngTransclude): don't overwrite the contents with an unfilled optional slot

Previously the contents of the `ngTransclude` element would always be emptied,
even if there was no transclusion to replace it.
Now, optional slots that have not been filled with content will not cause
the `ngTransclude` contents to be emptied.

Closes #13426

petebacondarwin commented:
Hmm it is a bit awkward to make the default transclusion optional. I think we should punt that feature to post-1.5.0

Narretz commented:
Looks like the last commit caused some doc e2e tests to fail. :o Other than that, feel free to merge. We can tweak it after the fact.

petebacondarwin committed:
docs($compile): add better multi-slot transclusion information

petebacondarwin committed:
fix($compile): swap keys and values for transclude definition object

Closes #13439

dmaslov committed:
fix(ngTransclude): don't replace existing content if no transcluded content exists

Closes #11839

petebacondarwin committed:
docs(ngTransclude): add better multi-slot translusion docs
